### ISSUE TITLE
mypy can not correctly tell when optional type can not be None

### ISSUE THREAD

**Issue by pbrezina** [python/mypy#14266]
**Bug Report**

mypy can not correctly tell when optional type can not be None if the variable was previously set or checked in @property.

**To Reproduce**

```python
class Test(object):
    def __init__(self) -> None:
        self.test: str | None = None

    @property
    def is_set(self) -> bool:
        return self.test is not None

    def failed_check(self) -> None:
        if not self.is_set:
            return

        # self.test can not be None
        value = self.test.strip(',')
        print(value)

class Test2(object):
    def __init__(self) -> None:
        self.test: str | None = None

    def set_me(self) -> None:
        self.test = 'test'

    def use_me(self) -> None:
        self.set_me()

        # self.test can not be None
        self.test.strip(',')
```

**Expected Behavior**

No issues found.

**Actual Behavior**

```
test.py:14: error: Item "None" of "Optional[str]" has no attribute "strip"  [union-attr]
test.py:28: error: Item "None" of "Optional[str]" has no attribute "strip"  [union-attr]
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: mypy 0.991 (compiled: yes)
- Mypy command-line flags: `mypy test.py`
- Mypy configuration options fro m `mypy.ini` (and other config files):
- Python version used: Python 3.10.8

**Comment by thomkeh** [comment]
Here is a work-around, but it's a bit ugly:

```python
from __future__ import annotations
from typing import Generic, Union, TypeGuard, TypeVar, cast

T = TypeVar("T", bound=Union[str, None])

def is_set(self: Test[T]) -> TypeGuard[Test[str]]:
    return self.test is not None

class Test(Generic[T]):
    test: T

    def __init__(self: Test[None]) -> None:
        self.test = None

    def failed_check(self) -> None:
        if is_set(self):
            reveal_type(self.test) # Revealed type is "builtins.str"
            value = self.test.strip(',')
            print(value)


def set_me(self: Test2[T]) -> TypeGuard[Test2[str]]:
    self_ = cast(Test2[str], self)
    self_.test = 'test'
    return True

class Test2(Generic[T]):
    test: T

    def __init__(self: Test2[None]) -> None:
        self.test = None

    def use_me(self) -> None:
        assert set_me(self)

        reveal_type(self.test) # Revealed type is "builtins.str"
        self.test.strip(',')
```

You can also play around with it here: https://mypy-play.net/?mypy=latest&python=3.11&gist=69c7abde15c48210f8c36c1a87a377f5

**Comment by pbrezina** [comment]
It is interesting workaround, thank you. But I'd rather ignore the error instead of using it. Is it something that can/will be fixed?

**Comment by mevol** [comment]
Is this one still an open issue? I have just run into the same problem that my variable can either be a string or None. For the latter I get the error:
error: Item "None" of "str | None" has no attribute "strip"  [union-attr]
